### ISSUE TITLE
Formatting with multi_json ?

### ISSUE THREAD

**Issue by ericproulx** [ruby-grape/grape#2486]
Am I wrong to say that [Grape::Formatter::JSON](https://github.com/ruby-grape/grape/blob/2b8567a1ca49b3d0236fa82deb2edef34c8839cf/lib/grape/formatter/json.rb#L8) will never call `Grape::Json::Dump` since every object responds to `:to_json` ?

**Comment by dblock** [comment]
We've relied on `multi_json` to swap JSON serializers between the default one and `oj`, so removing it would be backwards incompatible and definitely create problems for users. It's possible that the paths where it was meaningful were all with large amounts of JSON, so nobody noticed that the error formatter wasn't using it?

**Comment by ericproulx** [comment]
To code must call `Grape::Json::Dump` or `to_json` when formatting JSON ? The docs says
```
Built-in formatters are the following.
    :json: use object's to_json when available, otherwise call MultiJson.dump
```

This means that even if you have `multi_json`, the latter is not used when formatting JSON. `Grape::ErrorFormatter::Json` and `Grape::Parser::Json` are ok since its using `Grape::Json.dump` and `Grape::Json.load` respectively.

Its true that `Multijson` facilitates the usage of Oj but anybody could simply call `Oj.mimic_JSON()` and `Oj` would override all `to_json`.

**Comment by dblock** [comment]
> Its true that `Multijson` facilitates the usage of Oj but anybody could simply call `Oj.mimic_JSON()` and `Oj` would override all `to_json`.

It could be a relic of the past. A change like this is still breaking, so it's got to be worth it.

**Comment by ericproulx** [comment]
GitLab has their own [JSON Formatter](https://gitlab.com/gitlab-org/gitlab/-/blob/master/lib/gitlab/json.rb#L180) which is using `dump` instead of `to_json`. I think we should remove the `to_json` part and just keep the `Grape::Json.dump`.

**Comment by ericproulx** [comment]
Closing this as won't-fix. I implemented the change in #2769 (remove the `to_json` short-circuit so formatting always goes through `Grape::Json.dump`) and then closed that PR, because the premise behind it turned out to be wrong.

The concern here was that `Grape::Formatter::Json` "never reaches `Grape::Json.dump`" since every object responds to `to_json`, making the `multi_json` configuration inconsistent vs. `ErrorFormatter::Json` / `Parser::Json`. But that framed the short-circuit as a no-op that only bypasses the back-end. It isn't a no-op: `object.to_json` and `Grape::Json.dump(object)` produce **different output**.

`to_json` routes through each object's own JSON representation (ActiveSupport's `as_json`), while `Grape::Json.dump` (→ `JSON.generate`) only handles JSON primitives natively and falls back to `to_s` for everything else. So removing the short-circuit would silently change serialization for `Time`, `Date`, `BigDecimal`, symbols, and any object defining `as_json`/`to_json` — and can even flip the string encoding. For example:

```ruby
t = Time.at(0).getlocal("+00:00")

t.to_json            # => "1970-01-01T00:00:00.000+00:00"  (ISO 8601, honors as_json)
Grape::Json.dump(t)  # => "1970-01-01 00:00:00 +0000"      (Time#to_s fallback)
```

That's exactly the "it's breaking, so it's got to be worth it" bar @dblock raised — and it isn't worth it. The cost isn't just losing the configured back-end on that path; it's regressing established `as_json`/ISO 8601 output that users rely on (the common case with ActiveSupport loaded).

So the documented behavior — `:json: use object's to_json when available, otherwise call MultiJson.dump` — is intentional and stays. The `to_json` path deliberately honors per-object serialization; `multi_json`/`oj` users who want the back-end everywhere can use `Oj.mimic_JSON()` to override `to_json` globally.

`ErrorFormatter::Json` and `Parser::Json` remain on `Grape::Json` because they deal with Grape-controlled hashes/strings where there's no `as_json` semantics to preserve.

**Comment by ericproulx** [comment]
One clarification to be precise about *where* the divergence comes from, since it sharpens the rationale rather than weakening it.

Grape itself does **not** load ActiveSupport's JSON extension. It requires only a curated subset of `active_support/core_ext/*` (see `lib/grape.rb`), and `active_support/core_ext/object/json` — the file that defines `Object#as_json` and overrides `to_json` on `Time`/`Date`/`BigDecimal`/etc. — is **not** in that list (the railtie doesn't load it either).

So with `require "grape"` alone (non-Rails), `to_json` is the **stdlib** `json` gem's, and the two paths are actually equivalent:

```
# Grape only (no Rails)
t.to_json            # => "1970-01-01 00:00:00 +0000"
Grape::Json.dump(t)  # => "1970-01-01 00:00:00 +0000"   (equal)

# Grape + active_support/core_ext/object/json loaded (i.e. Rails)
t.to_json            # => "1970-01-01T00:00:00.000+00:00"   (ISO 8601, as_json)
Grape::Json.dump(t)  # => "1970-01-01 00:00:00 +0000"        (Time#to_s, differs)
```

In other words, the breaking divergence is introduced by the **host app** loading AS's JSON core extension — overwhelmingly **Rails apps**, which pull it in at boot. That's the majority of Grape's user base, and it's exactly those users who would silently regress from ISO 8601 to `to_s` output if the `to_json` short-circuit were removed. So keeping it stands.
